### ISSUE TITLE
Unexpected overhead when using `grad(pmap)`

### ISSUE THREAD

**Issue by izmailovpavel** [google/jax#3612]
When applying `grad` transformation to a function that uses `pmap`, it seems like unexpected compilation overhead is incurred. As an example, consider the following code (gradient computation in a linear regression problem with `80k` datapoints and `1k` features)
``` 
x = jax.random.normal(key=jax.random.PRNGKey(1), shape=(8, 1000, 10000))
y = jax.random.normal(key=jax.random.PRNGKey(2), shape=(8, 10000))
w = jax.random.normal(key=jax.random.PRNGKey(3), shape=(1000,))

def per_core_loss(w, x, y):
  loss_val = (w @ x - y)**2
  loss_val = jnp.mean(loss_val)
  return loss_val

@functools.partial(jax.pmap, axis_name='i')
def loss_p(w, x, y):
  loss_val, grad = jax.value_and_grad(per_core_loss)(w, x, y)
  loss_val = jax.lax.pmean(loss_val, axis_name="i")
  return loss_val

def loss_wrapper(w, x, y):
  w_p = jax.pmap(lambda _: w)(jnp.arange(8))
  loss_val = loss_p(w_p, x, y)[0]
  return loss_val

loss_val = loss_wrapper(w, x, y)
grad = jax.grad(loss_wrapper)(w, x, y)
```

Evaluating grad this way takes 18ms. At the same time, the following code that `pmaps` evaluation of `grad` runs in 9ms.

```
@functools.partial(jax.pmap, axis_name='i')
def loss_grad(w, x, y):
  loss_val, grad = jax.value_and_grad(per_core_loss)(w, x, y)
  loss_val = jax.lax.pmean(loss_val, axis_name="i")
  grad = jax.lax.pmean(grad, axis_name="i")
  return loss_val, grad

w_p = jax.pmap(lambda _: w)(jnp.arange(8))
loss_val, grad = loss_grad(w_p, x, y)
```
